torch/csrc/jit/fuser/compiler.cpp

```diff
@@ -144,6 +156,111 @@ static void setInputBroadcastGroups(KernelSpec& spec) {
+          // (i.e. the type)
+          JIT_ASSERT(offset == 1);
+          uses_to_process.insert(
+              uses_to_process.end(),
+              user->output()->uses().begin(),
+              user->output()->uses().end());
+        } else if (commutes_with_SumToSize.find(user)) {
+          // these are expressions that might occur during autotdiff operating
+          // on the gradient (matmul would likely be, too but we don't fuse it)
+          // note that for mul, we know (from the chain rule) that only one
+          // factor will be stemming from a calculation involving gradients so
+          // we know that we can move SumToSize across it
+          uses_to_process.insert(
+              uses_to_process.end(),
+              user->output()->uses().begin(),
+              user->output()->uses().end());
+        } else if (user->kind() == aten::_grad_sum_to_size) {
+          // this case should not happen due to how we organize the search
+          // and remove nodes
+          JIT_ASSERT(false);
+        } else if (user->kind() == prim::Return) {
+          // only if we don't already have a _grad_sum_to_size for this,
+          // because we want later (in the graph) nodes to be preferrable
+          // to earlier ones and we iterate backwards
+          // The implicit assumption is that if we have several _grad_sumtosizes
+          // "in parallel" (from addition) they are the same.
+          if (outputGradSumToSizes[offset] == -1) {
+            // note: we make the assumption that the sizes are inputs to the
+            // fusion group (rather than something calculated).
+            outputGradSumToSizes[offset] = node->inputs()[1]->offset();
+          }
+        } else if (user->kind() == prim::FusedConcat) {
+          throw std::runtime_error("don't know how to do fused concat yet");
+        } else {
+          throw std::runtime_error(
+              "unknown node type for fusion SumToSize processing");
+        }
+      }
+      // remove the GradSumToSize node, a new node outside the fusion graph
+      // will be inserted below
+      node->output()->replaceAllUsesWith(node->inputs()[0]);
+      it.destroyCurrent();
+    }
+  }
+  // By removing the _grad_sum_to_size notes, we might end up with
+  // duplicate outputs, e.g. when having the autodiff backwards of
+  // x + y + z of something with x, y, z, those will have different
+  // _grad_sum_to_sizes but of the same kernel output.
+  std::unordered_map<const Value*, int64_t> reduced_output_indices;
+  int64_t newo = 0;
+  for (auto osize : outputGradSumToSizes) {
+    auto it = reduced_output_indices.find(graph->outputs()[newo]);
+    if (it == reduced_output_indices.end()) {
+      reduced_output_indices.emplace(graph->outputs()[newo], newo);
+      outputMapAndSizes.emplace_back(newo, osize);
+      newo++;
+    } else {
+      graph->eraseOutput(newo);
+      outputMapAndSizes.emplace_back(it->second, osize);
+    }
+  }
```

Comment:
Ohhh interesting. So `outputMapAndSizes` really describes the postprocessing we do on the kernel outputs, and we might duplicate them! That should really go in the some comments.